File: parabolic/example/prep_grid/model1.m
% creat boundary data file bz1 bz2

clc;
clear all;
close all;

flag_printf = 1;
flag_topo_z = 1;

nx = 401;
ny = 401;

dx = 10;
dy = 10;

nz = 201;
dz = 10;

origin_x = 0;
origin_y = 0;
origin_z = 0;
bz1 = zeros(nx,ny,3);
bz2 = zeros(nx,ny,3);

for j=1:ny
  for i=1:nx
    bz2(i,j,1) = origin_x + (i-1)*dx;
    bz2(i,j,2) = origin_y + (j-1)*dy;
    bz2(i,j,3) = origin_z;
  end
end

if flag_topo_z
  x1=1000;
  y1=2000;
  x2=3000;
  y2=2000;
  L = 200;
  H = 500;
  for j = 1:ny
    for i = 1:nx
      r1 = (bz2(i,j,1)-x1)^2 + (bz2(i,j,2)-y1)^2;
      r2 = (bz2(i,j,1)-x2)^2 + (bz2(i,j,2)-y2)^2;

      topo = H * exp(-r1/L^2) - H * exp(-r2/L^2);
        
      bz2(i,j,3) = bz2(i,j,3) + topo;
    end
  end
end

% smooth alone two direction
for i=1:nx
 bz2(i,:,1) = smooth(bz2(i,:,1),20);
 bz2(i,:,2) = smooth(bz2(i,:,2),20);
 bz2(i,:,3) = smooth(bz2(i,:,3),20);
end

for j=1:ny
bz2(:,j,1) = smooth(bz2(:,j,1),20);
bz2(:,j,2) = smooth(bz2(:,j,2),20);
bz2(:,j,3) = smooth(bz2(:,j,3),20);
end

bz1(:,:,1) = bz2(:,:,1);
bz1(:,:,2) = bz2(:,:,2);
bz1(:,:,3) = -2000;

% A = 0.000001;
% [bz2] = arc_stretch_x(A,bz2);


figure(1)
surf(bz2(:,:,1)/1e3,bz2(:,:,2)/1e3,bz2(:,:,3)/1e3);
shading interp;
hold on;
plot3(bz2(:,201,1)/1e3,bz2(:,201,2)/1e3,bz2(:,201,3)/1e3,'r',LineWidth=2);

plot3(1.0, 1.0, +0.1, 'rv',MarkerSize=20,LineWidth=2);
plot3(1.5, 1.0, +0.1, 'rv',MarkerSize=20,LineWidth=2);
plot3(2.0, 1.0, +0.1, 'rv',MarkerSize=20,LineWidth=2);
plot3(2.5, 1.0, +0.1, 'rv',MarkerSize=20,LineWidth=2);
plot3(3.0, 1.0, +0.1, 'rv',MarkerSize=20,LineWidth=2);
text(1.0, 1.0-0.1, +0.1+0.2,'R1',Color='r',FontSize=20,FontWeight='bold');
text(1.5, 1.0-0.1, +0.1+0.2,'R2',Color='r',FontSize=20,FontWeight='bold');
text(2.0, 1.0-0.1, +0.1+0.2,'R3',Color='r',FontSize=20,FontWeight='bold');
text(2.5, 1.0-0.1, +0.1+0.2,'R4',Color='r',FontSize=20,FontWeight='bold');
text(3.0, 1.0-0.1, +0.1+0.2,'R5',Color='r',FontSize=20,FontWeight='bold');
plot3(3.0,2.5,0.1,'rp',MarkerSize=20,LineWidth=2);

grid off;
axis equal;
%caxis([-0.5,0]);
cid = colorbar;
cid.Label.String='(km)';
set(gca,'layer','top');
set(gcf,'color','white');
set(gcf,'Position',[0,0,2200,1200]);
set(gca, FontWeight='bold', FontSize=15);
xlabel('X axis (km)',FontWeight='bold',rotation=-25,position=[2.7 -1.2]);
ylabel('Y axis (km)',FontWeight='bold',rotation=35, position=[5.1 1.0]);
zlabel('Z axis (km)',FontWeight='bold',rotation=0,position=[-0.6 0.7]);
view(40,30);

grid off;
% creat data file
file_name = '../data_file_3d.txt';
fid=fopen(file_name,'w'); % Output file name 
fprintf(fid,'# nx ny number\n'); 
fprintf(fid,'%d %d\n',nx,ny);
fprintf(fid,'# bz1 coords\n'); 
for j=1:ny
  for i=1:nx
    fprintf(fid,'%.9e %.9e %.9e\n',bz1(i,j,1),bz1(i,j,2),bz1(i,j,3));
  end
end
fprintf(fid,'# bz2 coords\n'); 
for j=1:ny
  for i=1:nx
    fprintf(fid,'%.9e %.9e %.9e\n',bz2(i,j,1),bz2(i,j,2),bz2(i,j,3));
  end
end


if flag_printf
   print(gcf,'model1.png','-r300','-dpng');
end

File: parabolic/example/prep_grid/model1.py
import sys, os
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', '..', '..', 'plotting'))
from export_c import export_bdry_2

import numpy as np
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt


def smooth_1d(arr, span):
    """Reimplementation of MATLAB smooth(y, span) with 'moving' method.

    Even span is reduced by 1 (MATLAB convention). At boundaries, uses
    anchored windows (left-anchored at start, right-anchored at end)
    rather than centered truncated windows.
    """
    arr = np.asarray(arr, dtype=float)
    n = len(arr)
    if span % 2 == 0:
        span = span - 1
    half = span // 2
    result = np.zeros(n)
    for i in range(n):
        if i < half:
            lo = 0
            hi = 2 * i
        elif i >= n - half:
            lo = 2 * i - n + 1
            hi = n - 1
        else:
            lo = i - half
            hi = i + half
        result[i] = np.mean(arr[lo:hi+1])
    return result


def main():
    flag_printf = 1
    flag_topo_z = 1

    nx = 401
    ny = 401
    dx = 10
    dy = 10

    origin_x = 0
    origin_y = 0
    origin_z = 0

    bz1 = np.zeros((nx, ny, 3))
    bz2 = np.zeros((nx, ny, 3))

    for j in range(ny):
        for i in range(nx):
            bz2[i, j, 0] = origin_x + i * dx
            bz2[i, j, 1] = origin_y + j * dy
            bz2[i, j, 2] = origin_z

    if flag_topo_z:
        x1 = 1000
        y1 = 2000
        x2 = 3000
        y2 = 2000
        L = 200
        H = 500
        for j in range(ny):
            for i in range(nx):
                r1 = (bz2[i, j, 0] - x1)**2 + (bz2[i, j, 1] - y1)**2
                r2 = (bz2[i, j, 0] - x2)**2 + (bz2[i, j, 1] - y2)**2
                topo = H * np.exp(-r1 / L**2) - H * np.exp(-r2 / L**2)
                bz2[i, j, 2] += topo

    # Smooth along two directions
    window_len = 20
    for i in range(nx):
        bz2[i, :, 0] = smooth_1d(bz2[i, :, 0], window_len)
        bz2[i, :, 1] = smooth_1d(bz2[i, :, 1], window_len)
        bz2[i, :, 2] = smooth_1d(bz2[i, :, 2], window_len)
    for j in range(ny):
        bz2[:, j, 0] = smooth_1d(bz2[:, j, 0], window_len)
        bz2[:, j, 1] = smooth_1d(bz2[:, j, 1], window_len)
        bz2[:, j, 2] = smooth_1d(bz2[:, j, 2], window_len)

    # bz1 gets x,y from bz2, z=-2000
    bz1[:, :, 0] = bz2[:, :, 0]
    bz1[:, :, 1] = bz2[:, :, 1]
    bz1[:, :, 2] = -2000

    file_name = '../data_file_3d.txt'
    export_bdry_2(bz1, bz2, nx, ny, file_name)

    if flag_printf:
        os.makedirs('../plot', exist_ok=True)
        stride = max(1, max(nx, ny) // 80)
        fig = plt.figure(figsize=(11, 6))
        ax = fig.add_subplot(111, projection='3d')
        X = bz2[:, :, 0] / 1e3; Y = bz2[:, :, 1] / 1e3; Z = bz2[:, :, 2] / 1e3
        ax.plot_surface(X[::stride, ::stride], Y[::stride, ::stride],
                        Z[::stride, ::stride], cmap='jet', alpha=0.9, edgecolor='none')
        mid_j = ny // 2
        ax.plot(X[:, mid_j], Y[:, mid_j], Z[:, mid_j], 'r', linewidth=2)
        ax.set_xlabel('X axis (km)', fontweight='bold')
        ax.set_ylabel('Y axis (km)', fontweight='bold')
        ax.set_zlabel('Z axis (km)', fontweight='bold')
        ax.view_init(elev=30, azim=40)
        ax.set_box_aspect([1, 1, 0.5])
        ax.grid(False)
        mappable = plt.cm.ScalarMappable(cmap='jet')
        mappable.set_array(Z[::stride, ::stride].ravel())
        mappable.set_clim(Z.min(), Z.max())
        cb = fig.colorbar(mappable, ax=ax, shrink=0.6, pad=0.1)
        cb.set_label('(km)', fontweight='bold')
        fig.patch.set_facecolor('white')
        plt.tight_layout()
        plt.savefig('../plot/para_model1.png', dpi=300, bbox_inches='tight', facecolor='white')
        plt.close(fig)
        print('Saved: ../plot/para_model1.png')


if __name__ == "__main__":
    main()
